geometry/instant_output_test.cpp

```diff
@@ -0,0 +1,141 @@
+﻿
+#include <strstream>
+
+#include "astronomy/time_scales.hpp"
+#include "geometry/named_quantities.hpp"
+#include "geometry/sign.hpp"
+#include "gtest/gtest.h"
+#include "gmock/gmock.h"
+
+namespace principia {
+namespace geometry {
+namespace internal_point {
+
+using astronomy::InfiniteFuture;
+using astronomy::InfinitePast;
+using astronomy::J2000;
+using astronomy::operator""_TT;
+using astronomy::operator""_UT1;
+using astronomy::operator""_UTC;
+using quantities::Time;
+using quantities::si::Second;
+using ::testing::Eq;
+
+class InstantOutputTest : public ::testing::Test {
+ protected:
+  Instant JustAfter(Instant const t) {
+    return J2000 + std::bit_cast<double>(
+                       std::bit_cast<std::uint64_t>((t - J2000) / Second) +
+                       Sign(t - J2000)) *
+                       Second;
+  }
+  Instant JustBefore(Instant const t) {
+    return J2000 + std::bit_cast<double>(
+                       std::bit_cast<std::uint64_t>((t - J2000) / Second) -
+                       Sign(t - J2000)) *
+                       Second;
+  }
+
+  static constexpr Instant negative_j2000_{-0.0 * Second};
+};
+
+TEST_F(InstantOutputTest, UniversalTime) {
+  // Note that since we are roughly twelve hours from J2000, we expect
+  // log₁₀(12 h / 1 s), between four and five, of the printed digits to be
+  // unrepresentable: our granularity is about seven picoseconds. This is
+  // obvious in the case of UTC, which is offset from TT(TAI) by an integer
+  // number of milliseconds at any time.
+  EXPECT_THAT((std::stringstream() << "2000-01-01T00:00:00"_UTC).str(),
+              Eq("2000-01-01T00:01:04,1840000000011059 (TT)"));
```

Comment:
The trailing `11059` is not significant, right?  It might change if the implementation changes.  I would use a regexp and leave these digits unspecified.

Reply:
The implementation (of operator""_UTC) should be correctly-rounded (it is not, see the TODO below, this was found by looking at those « insignificant » digits, so they test _something_).
Barring changes of the value being formatted, these digits must not change: the fractional part (stored in `remainder`) is computed exactly, and we print 17 digits of it; if we cannot trust the binary-to-decimal code on seventeen digits we cannot ever hope that it will survive round-tripping.